asm/vm/asm_vm

```diff
@@ -677,27 +836,62 @@ retrieve_cluster_root_cert() {
   info "Retrieving the cluster's root certificate..."
 
   local ISTIOD; ISTIOD=""
-  ISTIOD=$(kubectl get pods -oname -lapp=istiod -n istio-system | head -n 1 || true)
-  ROOT_CERT=$(kubectl exec -it -n istio-system "${ISTIOD}" \
+  ISTIOD="$(retry 2 kubectl get pods -oname -lapp=istiod -n istio-system | head -n 1)"
+  ROOT_CERT="$(retry 2 run kubectl exec -it -n istio-system "${ISTIOD}" \
     -- cat /var/run/secrets/kubernetes.io/serviceaccount/ca.crt \
-    | awk 'NF {sub(/\r/, ""); printf "%s\n",$0;}' || true)
+    | awk 'NF {sub(/\r/, ""); printf "%s\n",$0;}')"
   if [[ -z "${ROOT_CERT}" ]]; then
     fatal "Error retrieving the cluster's root certificate from Istiod pod ${ISTIOD}. Please try again."
   fi
 }
 
-retrieve_ingressgateway() {
-  info "Retrieving the cluster's Istio ingress gateway IP..."
+retrieve_eastwestgateway() {
+  info "Retrieving the cluster's Istio eastwest gateway IP..."
 
-  local RETVAL
-  { INGRESS_IP=$(watch_for_result 15 "${IP_REGEX}" kubectl get service \
-    istio-ingressgateway -n istio-system \
-    -o jsonpath='{.status.loadBalancer.ingress[0].ip}'); RETVAL="$?"; } || true
-  if [[ "${RETVAL}" -ne 0 ]]; then
-    fatal "Failed to retrieve the Istio ingress gateway IP address."
+  INGRESS_IP="$(retry 2 run kubectl get service "${EXPANSION_GATEWAY_NAME}" \
+    -n istio-system -o jsonpath='{.status.loadBalancer.ingress[0].ip}')"
+  if [[ -z "${INGRESS_IP}" ]]; then
+    fatal "Failed to retrieve the Istio eastwest gateway IP address. Please try again."
   fi
 }
 
+retrieve_cluster_network() {
+  info "Retrieving the cluster's network name..."
+
+  NETWORK="$(retry 2 run gcloud container clusters describe "${CLUSTER_NAME}" \
+    --zone="${CLUSTER_LOCATION}" \
+    --project="${PROJECT_ID}" \
+    --format="value(network)")"
+}
+
+retrieve_identity_provider() {
+  retrieve_membership_uid
+
+  CREDENTIAL_IDENTITY_PROVIDER="${MEMBERSHIP_UID}@${WORKLOAD_IDENTITY_PROVIDER}@${WORKLOAD_NAMESPACE}"
+}
+
+retrieve_membership_uid() {
+  info "Retrieving the Environ membership's unique ID..."
+
+  local MEMBERSHIPS
+  MEMBERSHIPS="$(retry 2 gcloud container hub memberships list --project "${PROJECT_ID}" \
+   --format='value(name)')"
+  while read -r MEMBERSHIP; do
+    if [[ -n "${MEMBERSHIP}" ]]; then
+      local CURR_CLUSTER_URL
+      CURR_CLUSTER_URL="$(retry 2 gcloud container hub memberships describe \
+        "${MEMBERSHIP}" --project "${PROJECT_ID}" \
+        --format="value(endpoint.gkeCluster.resourceLink)")"
+      if [[ "${CURR_CLUSTER_URL}" == "${GKE_URL}" ]]; then
+        MEMBERSHIP_UID="$(retry 2 run gcloud container hub memberships describe \
+          "${MEMBERSHIP}" --project "${PROJECT_ID}" --format="value(uniqueId)")"
+      fi
+    fi
+  done <<EOF
+${MEMBERSHIPS}
+EOF
+}
+
 # Generate GCE service proxy definition for the instance template
```

Comment:
You don't break after you find what you're looking for. It looks like you might find multiple UIDs but only set the latest one. Can you return early? Also, is there a reason not to just set CREDENTIAL_IDENTITY_PROVIDER in the same function?

Reply:
Good point. break is added. But it will not have more than one UIDs returned, break is making it more efficient.

> Also, is there a reason not to just set CREDENTIAL_IDENTITY_PROVIDER in the same function?

Not really, just separation of logic. But combined.